TestIcebergSmokeRest AbstractTestIntegrationSmokeTest.testSelectInformationSchemaColumns falked

Error:  Tests run: 2916, Failures: 1, Errors: 0, Skipped: 34, Time elapsed: 2,342.931 s <<< FAILURE! - in TestSuite
Error:  com.facebook.presto.iceberg.rest.TestIcebergSmokeRest.testSelectInformationSchemaColumns  Time elapsed: 0.117 s  <<< FAILURE!
java.lang.AssertionError: Execution of 'actual' query failed: SELECT table_schema FROM information_schema.columns WHERE table_schema = 'tpch' GROUP BY table_schema
	at org.testng.Assert.fail(Assert.java:98)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:178)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:106)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:164)
	at com.facebook.presto.tests.AbstractTestIntegrationSmokeTest.testSelectInformationSchemaColumns(AbstractTestIntegrationSmokeTest.java:162)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.RuntimeException: Table does not exist: tpch.test_truncate_transform
	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:124)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:733)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:701)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:175)
	... 18 more
Caused by: org.apache.iceberg.exceptions.NoSuchTableException: Table does not exist: tpch.test_truncate_transform
	at org.apache.iceberg.rest.ErrorHandlers$TableErrorHandler.accept(ErrorHandlers.java:112)
	at org.apache.iceberg.rest.ErrorHandlers$TableErrorHandler.accept(ErrorHandlers.java:102)
	at org.apache.iceberg.rest.HTTPClient.throwFailure(HTTPClient.java:201)
	at org.apache.iceberg.rest.HTTPClient.execute(HTTPClient.java:313)
	at org.apache.iceberg.rest.HTTPClient.execute(HTTPClient.java:252)
	at org.apache.iceberg.rest.HTTPClient.get(HTTPClient.java:348)
	at org.apache.iceberg.rest.RESTClient.get(RESTClient.java:96)
	at org.apache.iceberg.rest.RESTSessionCatalog.loadInternal(RESTSessionCatalog.java:331)
	at org.apache.iceberg.rest.RESTSessionCatalog.loadTable(RESTSessionCatalog.java:347)
	at org.apache.iceberg.catalog.BaseSessionCatalog$AsCatalog.loadTable(BaseSessionCatalog.java:99)
	at org.apache.iceberg.rest.RESTCatalog.loadTable(RESTCatalog.java:102)
	at com.facebook.presto.iceberg.IcebergUtil.getNativeIcebergTable(IcebergUtil.java:230)
	at com.facebook.presto.iceberg.IcebergNativeMetadata.getRawIcebergTable(IcebergNativeMetadata.java:91)
	at com.facebook.presto.iceberg.IcebergAbstractMetadata.lambda$getIcebergTable$0(IcebergAbstractMetadata.java:192)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1688)
	at com.facebook.presto.iceberg.IcebergAbstractMetadata.getIcebergTable(IcebergAbstractMetadata.java:190)
	at com.facebook.presto.iceberg.IcebergAbstractMetadata.getTableMetadata(IcebergAbstractMetadata.java:369)
	at com.facebook.presto.iceberg.IcebergAbstractMetadata.listTableColumns(IcebergAbstractMetadata.java:392)
	at com.facebook.presto.spi.connector.classloader.ClassLoaderSafeConnectorMetadata.listTableColumns(ClassLoaderSafeConnectorMetadata.java:338)
	at com.facebook.presto.metadata.MetadataManager.listTableColumns(MetadataManager.java:574)
	at com.facebook.presto.metadata.MetadataListing.listTableColumns(MetadataListing.java:95)
	at com.facebook.presto.connector.informationSchema.InformationSchemaPageSourceProvider.buildColumns(InformationSchemaPageSourceProvider.java:153)
	at com.facebook.presto.connector.informationSchema.InformationSchemaPageSourceProvider.getInformationSchemaTable(InformationSchemaPageSourceProvider.java:122)
	at com.facebook.presto.connector.informationSchema.InformationSchemaPageSourceProvider.getInternalTable(InformationSchemaPageSourceProvider.java:116)
	at com.facebook.presto.connector.informationSchema.InformationSchemaPageSourceProvider.createPageSource(InformationSchemaPageSourceProvider.java:86)
	at com.facebook.presto.spi.connector.ConnectorPageSourceProvider.createPageSource(ConnectorPageSourceProvider.java:52)
	at com.facebook.presto.split.PageSourceManager.createPageSource(PageSourceManager.java:80)
	at com.facebook.presto.operator.ScanFilterAndProjectOperator.getOutput(ScanFilterAndProjectOperator.java:250)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:441)
	at com.facebook.presto.operator.Driver.lambda$processFor$10(Driver.java:324)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:750)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:317)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1079)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:165)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:621)
	at com.facebook.presto.$gen.Presto_0_288_SNAPSHOT_982bd09__testversion____20240529_181412_4.run(Unknown Source)
	... 3 more

[INFO] 
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:    TestIcebergSmokeRest>AbstractTestIntegrationSmokeTest.testSelectInformationSchemaColumns:162->AbstractTestQueryFramework.assertQuery:164 Execution of 'actual' query failed: SELECT table_schema FROM information_schema.columns WHERE table_schema = 'tpch' GROUP BY table_schema
[INFO] 
Error:  Tests run: 2916, Failures: 1, Errors: 0, Skipped: 34

--- a/presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java
+++ b/presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java
@@ -86,6 +86,7 @@ import org.apache.iceberg.SchemaParser;
 import org.apache.iceberg.Table;
 import org.apache.iceberg.Transaction;
 import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.exceptions.NoSuchViewException;
 import org.apache.iceberg.types.Type;
 import org.apache.iceberg.types.TypeUtil;
 import org.apache.iceberg.types.Types;
@@ -396,8 +397,9 @@ public abstract class IcebergAbstractMetadata
 
     protected ConnectorTableMetadata getTableOrViewMetadata(ConnectorSession session, SchemaTableName table, IcebergTableName icebergTableName)
     {
+        SchemaTableName schemaTableName = new SchemaTableName(table.getSchemaName(), icebergTableName.getTableName());
         try {
-            Table icebergTable = getIcebergTable(session, new SchemaTableName(table.getSchemaName(), icebergTableName.getTableName()));
+            Table icebergTable = getIcebergTable(session, schemaTableName);
             ImmutableList.Builder<ColumnMetadata> columns = ImmutableList.builder();
             columns.addAll(getColumnMetadata(icebergTable));
             if (icebergTableName.getTableType() == CHANGELOG) {
@@ -409,12 +411,17 @@ public abstract class IcebergAbstractMetadata
             }
             return new ConnectorTableMetadata(table, columns.build(), createMetadataProperties(icebergTable), getTableComment(icebergTable));
         }
-        catch (NoSuchTableException e) {
+        catch (NoSuchTableException noSuchTableException) {
             // Considering that the Iceberg library does not provide an efficient way to determine whether
             //  it's a view or a table without loading it, we first try to load it as a table directly, and then
             //  try to load it as a view when getting an `NoSuchTableException`. This will be more efficient.
-            View icebergView = getIcebergView(session, new SchemaTableName(table.getSchemaName(), icebergTableName.getTableName()));
-            return new ConnectorTableMetadata(table, getColumnMetadata(icebergView), createViewMetadataProperties(icebergView), getViewComment(icebergView));
+            try {
+                View icebergView = getIcebergView(session, schemaTableName);
+                return new ConnectorTableMetadata(table, getColumnMetadata(icebergView), createViewMetadataProperties(icebergView), getViewComment(icebergView));
+            }
+            catch (NoSuchViewException noSuchViewException) {
+                throw new TableNotFoundException(schemaTableName);
+            }
         }
     }
 

